File: THCS/2022_2023/NgheAn_2223/Cau2/HangCay.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

int main() {
    std::ifstream input("HangCay.INP");
    std::ofstream output("HangCay.OUT");

    int n, m;
    input >> n >> m; // Đọc số cây và số vòi nước

    // Mảng để theo dõi các cây đã được tưới nước
    std::vector<bool> trees_watered(n + 1, false);

    // Đọc thông tin vòi nước
    for (int i = 0; i < m; ++i) {
        int x_i, r_i;
        input >> x_i >> r_i; // Đọc vị trí và bán kính vòi nước

        // Tính khoảng cách tưới nước
        int left_bound = std::max(1, x_i - r_i);
        int right_bound = std::min(n, x_i + r_i);

        // Đánh dấu các cây được tưới nước
        for (int j = left_bound; j <= right_bound; ++j) {
            trees_watered[j] = true;
        }
    }

    // Đếm số lượng cây đã được tưới nước
    int watered_count = 0;
    for (int i = 1; i <= n; ++i) {
        if (trees_watered[i]) {
            ++watered_count;
        }
    }

    // Xuất kết quả ra tệp
    output << watered_count << std::endl;

    input.close();
    output.close();

    return 0;
}

File: THCS/2022_2023/NgheAn_2223/Cau2/HangCay.py
def main():
    # Đọc dữ liệu từ tệp HangCay.INP
    with open('HangCay.INP', 'r') as input_file:
        n, m = map(int, input_file.readline().strip().split())  # Đọc n và m
        trees_watered = [False] * (n + 1)  # Mảng đánh dấu các cây được tưới nước

        # Đọc thông tin vòi nước
        for _ in range(m):
            x_i, r_i = map(int, input_file.readline().strip().split())  # Đọc vị trí và bán kính vòi nước
            # Tưới nước cho các cây trong khoảng [x_i - r_i, x_i + r_i]
            left_bound = max(1, x_i - r_i)
            right_bound = min(n, x_i + r_i)
            for j in range(left_bound, right_bound + 1):
                trees_watered[j] = True  # Đánh dấu cây được tưới nước

    # Tính số lượng cây được tưới nước
    watered_count = sum(trees_watered)  # Đếm số cây được tưới nước

    # Xuất kết quả ra tệp HangCay.OUT
    with open('HangCay.OUT', 'w') as output_file:
        output_file.write(str(watered_count) + '\n')

if __name__ == '__main__':
    main()
